proposals/sycl_modules.md

```diff
@@ -707,66 +738,186 @@ value to the specialization constant.
 
 ```cpp
 namespace sycl {
-namespace opencl {
-  template<typename ID, typename T>
-  spec_constant<ID, T> set_spec_constant(module<module_state::executable> syclModule, T cst) const;
-}  // namespace opencl
+
+template<class T, spec_id<T>& s>
+void set_spec_constant(module<module_status::input> syclModule, T);
+
+}  // namespace sycl
+```
+If C++17 or later is enabled, then the following overload is available
+```cpp
+namespace sycl {
+
+template<auto& s>
+void set_spec_constant(module<module_status::input> syclModule,
+                       typename std::remove_reference_t<decltype(s)>::type);
+
+}  // namespace sycl
+```
+
+It is valid for a user to set several times a spec constant with the same `spec_id`,
+only the last version of the value will be consider when building the module.
+Calling this function is not allowed during the submission of a task.
+
+### Getting the value of spec constant in kernel
+
+To retrive the value of a spec constant in a SYCL kernel context, the following API is available.
+
+```cpp
+namespace sycl {
+
+template<class T, spec_id<T>& s>
+T get_spec_constant(const spec_constant<T, s>&);
+
+}  // namespace sycl
+```
+If C++17 or later is enabled, then the following overload is available
+```cpp
+namespace sycl {
+
+template<auto& s>
+typename std::remove_reference_t<decltype(s)>::type
+get_spec_constant(const spec_constant<typename std::remove_reference_t<decltype(s)>::type, s>&);
+
 }  // namespace sycl
 ```
 
-For simplicity, the spec constant placeholder type can also be obtained from
-the command group handler:
+This call is only valid whitin a SYCL kernel context and the `spec_constant` value must have been
+created by the handler that invoked the kernel from which the `sycl::get_spec_constant` is called.
+
+If the module has native support for specialization constant,
+the returned value becomes a constant value available for the runtime compiler to use for optimisations.
+
+**Note:** the value may become a constant (known numerical value), although this may only happen at runtime.
+From the SYCL compiler's perspective (C++ compiler) the value is understood to be a runtime value and
+for this reason it is not possible to use such value in a `constexpr`.
+
+### Retrieving a spec constant placeholder
+
+A `spec_constant` handler can be retrieved from the command group handler by either:
+ - Getting the placeholder for a previously set specialization constant value;
+   - If no specialization constant where ever set, then the *placeholder* object holds the default value of its `spec_id`.
+ - Setting a specialization constant value and retrieving the placeholder
+   - Performance notice: this function may cause the SYCL runtime to implicitly recompile the relevant SYCL module with the given value. This may have an important effect on performances and users should check their vendor recommendations on how to efficiently uses this functionality.
```

Comment:
You can say you don't support the feature, or rely on a similar strategy as the host one.

I added this note it may not be obvious what is happening under the hood. I expect implementations to cache generated kernels, but if the user changes the value too often, there is nothing the runtime can really do and this will create a significant slowdown.

As this function is the most convenient one, I think it is important to mention it.